### PR TITLE
Clone-able signer

### PR DESCRIPTION
**Pull request by ltseeley**

### CODE REVIEW

File: libcylinder/src/signing/secp256k1/mod.rs [hunk 1]
```diff
@@ -52,30 +49,85 @@ impl Default for Secp256k1Context {
...
+        Ok(PublicKey::new(
+            secp256k1::key::PublicKey::from_secret_key(&self.context, &sk)
+                .serialize()
+                .to_vec(),
+        ))
+    }
+}
+
+#[derive(Clone)]
+pub struct Secp256k1Signer {
```
Comment (vaporos) [inline libcylinder/src/signing/secp256k1/mod.rs:77]: Does this an verifier impl need to be pub?

Comment (ltseeley) [reply libcylinder/src/signing/secp256k1/mod.rs]: They don't; I updated them to be private, we can always expose them at a later time if we have a need to.

File: libcylinder/tests/mod.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * ------------------------------------------------------------------------------
+ */
+
+use std::thread;
+
+use cylinder::{Context, PrivateKey, Signature};
+
+type TestResult = Result<(), Box<dyn std::error::Error>>;
+
+fn test_multithreaded_signing<C: Context + 'static>(context: C) -> TestResult {
```
Comment (vaporos) [inline libcylinder/tests/mod.rs:24]: Needs test doc header

File: libcylinder/tests/mod.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    let private_key = PrivateKey::new_from_hex(
+        "2f1e7b7a130d7ba9da0068b3bb0ba1d79e7e77110302c9f746c3c2a63fe40088",
```
Comment (vaporos) [inline libcylinder/tests/mod.rs:26]: This is secp256k1-specific, so should probably be passed into the fn.

File: libcylinder/src/signing/secp256k1/mod.rs [hunk 4]
```diff
@@ -153,46 +205,26 @@ mod secp256k1_test {
...
-        assert_eq!(priv_key.as_hex(), KEY1_PRIV_HEX);
-
-        ContextSigner::new_boxed(context, priv_key)
-    }
-
     #[test]
     fn single_key_signing_return_from_func() {
-        let signer = create_signer();
+        let key = PrivateKey::new_from_hex(KEY1_PRIV_HEX).expect("Failed to parse key from hex");
+        let signer = Secp256k1Context::new().new_signer(key);
```
[the remark this answers is not in the document]
Comment (ltseeley) [reply libcylinder/src/signing/secp256k1/mod.rs]: Fixed

File: libcylinder/tests/mod.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    let public_key = signer.public_key()?;
+    let verifier = context.new_verifier();
+    assert!(verifier.verify(b"Hello", &sig1, &public_key)?);
+    assert!(verifier.verify(b"Hello", &sig2, &public_key)?);
+
+    Ok(())
+}
+
+// TODO: test get signer from string?
```
[the remark this answers is not in the document]
Comment (ltseeley) [reply libcylinder/tests/mod.rs]: Removed the comment

File: libcylinder/tests/mod.rs (hunk 2 above)
Comment (ltseeley) [reply libcylinder/tests/mod.rs]: Added

File: libcylinder/tests/mod.rs (hunk 3 above)
Comment (ltseeley) [reply libcylinder/tests/mod.rs]: Fixed

**Pull request merged**